enclave/core/globals.c

```diff
@@ -8,6 +8,128 @@
+ * flags (-fPIE or -fpie) instead of position independent code flags
+ * (-fPIC or -fpic), then the variables defined in the compilation unit are
+ * to be given preference over definitions of the same variables in shared
+ * libraries. Hence when compiled using -fPIE/-fpie, local globals
+ * (type 2 above) are also guaranteed to be generated using RIP-relative
+ * addressing just like static globals.
+ *
+ *****************************************************************************
+ * Calculating the enclave base address securely
+ *
+ * The enclave can be loaded at any address in memory. Only the host knows
+ * the address where the enclave is loaded. So how can an enclave know its
+ * base address?
+ *
+ * Strategy 1:
+ *    Let the host set a variable oe_enclave_base within the enclave after
+ *    loading. This approach is not safe since the host is untrusted and could
+ *    set an arbitrary value and trick the enclave to writing secrets to host
+ *    memory.
+ *
+ * Strategy 2:
+ *    Define a variable called oe_enclave_rva (relative virtual address).
+ *    The oe_enclave_rva is supposed to be set to the offset of that variable
+ *    within the enclave image (i.e. as if the enclave was loaded to memory
+ *    address 0). This offset value is a constant and therefore the enclave
+ *    is signed after setting oe_enclave_rva to this value.
+ *
+ *    Actual runtime address of oe_enclave_rva minus the predetermined offset
+ *    of oe_enclave_rva gives the runtime base address of the enclave:
+ *       enclave-base-address = &oe_enclave_rva - oe_enclave_rva;
+ *
+ *    If the host sets oe_enclave_rva to an arbitrary value, then the loading
+ *    the enclave will fail since the signature (generated with incorrect
+ *    oe_enclave_rva value) will not match the signature embedded by the
+ *    oesign tool (using the correct oe_enclave_rva value).
+ *
+  *****************************************************************************
+ * Applying relocations
+ *
+ * Relocations are done within the enclave and relocation entries are measured
+ * as part of enclave signing. This makes relocations secure.
+ * Performing relocations involves adding the enclave base address to
+ * each relocation entry (see reloc.c in linux/windows).
+ * The Global Offset Table (GOT) is updated via performing the relocations.
+ * This means that the oe_enclave_rva variable and other variables used for
+ * relocations must not themselves be placed in the GOT. The simplest
+ * (though not the only) way to enforce RIP-relative addressing is to
+ * put all the relocation related variables into static global struct.
+ *
+ * Since the relocation values are supposed to be set by the loader,
+ * unbeknownst to the compiler, they must be volatile qualified to prevent
+ * compiler optimizations.
+ *
+ **/
+
+static volatile uint64_t oe_enclave_rva;
+static volatile uint64_t oe_reloc_rva;
+static volatile uint64_t oe_reloc_size;
+
```

Comment:
static naming conventions?

Reply:
I filed a new issue #1189  to update doc and fix the code to conform to the static variable naming conventions. @anakrish - But great if you can modify this before your checkin.

Reply:
Fix conventions in this PR.